### DISCUSSION TITLE
When will using <Suspense /> with Next.js be supported?

### DISCUSSION THREAD

**Discussion by hmbrg** [trpc/trpc#1677] [Q&A] [answered]
Next.js recently added support for React 18 and with that the new Streaming SSR API. I've tried it using the experimental flag in combination with setting

```js
      queryClientConfig: {
        defaultOptions: {
          queries: { suspense: true, useErrorBoundary: true },
          mutations: { useErrorBoundary: true },
        },
      },
```

in my `withTRPC()` call. 

This triggers the following error: `Error: ReactDOMServer does not yet support Suspense.`

Can this be fixed? Thanks in advance, looking forward to your replies!

**Comment by KATT** [comment]
This particular error doesn't look trpc related and rather that it isn't setup correctly. That said, trpc doesn't work with React 18 server components yet. 

I tried the nextjs RSC examples recently with the intention of making trpc support, but they were really flaky. It will probably come before Next.js officially supports React 18 and RSC.

If they make some announcement of moving to beta or RC, feel free to let me know, at the time I had a stab at it felt way too premature.

**Reply by statusunknown418** [reply]
Hey, have this got any update?

**Reply by KATT** [reply]
[There's hope](https://nextjs.org/blog/layouts-rfc)

**Comment by KATT** [comment, marked as answer]
https://trpc.io/docs/suspense

**Comment by dmaulick** [comment]
First of all I want to say I just started using this lib and am really enjoying it - I am using it to iterate very fast on both mobile and web like never before. So great work!

My question: 
What is the latest in terms of trpc <> suspense stability. I would like to migrate everything on web (next) and mobile (expo, may eject to rn soon) to suspense and not look back.  BUT I want to confirm that suspense support is a priority for the team. I am also open to supporting the effort if y'all have a thread on it already.

**Reply by KATT** [reply]
We support Suspense already and if you don't need SSR you can use it fine. I can't speak for RN though

- #3297
- https://trpc.io/docs/suspense

**Reply by hmbrg** [reply]
For now, we are using a polyfill until SSR & Suspense are fully supported. It looks something like this and should be easy to switch out with the actual Suspense component when it's ready.

```javascript
import { QueryFilters } from 'react-query/types/core/utils';

type SafeSuspenseProps = {
  children?: ReactNode;
  fallback?: ReactNode;
  selector?: QueryFilters;
};

const SafeSuspense: React.FC<SafeSuspenseProps> = ({
  children,
  fallback,
  selector,
}) => {
  return (
    <>
      {children}
      {useIsFetching(selector) > 0 ? (
        <div
          css={{
            position: 'absolute',
            top: '0px',
            bottom: '0px',
            left: '0px',
            right: '0px',
            zIndex: '100',
          }}
        >
          {fallback}
        </div>
      ) : null}
    </>
  );
};

export const SuspenseBoundary: React.FC<SafeSuspenseProps> = ({
  children,
  selector,
  fallback,
}) => {
  return (
    <DefaultErrorBoundary>
      <SafeSuspense
        selector={selector}
        fallback={fallback ?? <FallbackSuspense />}
      >
        {children}
      </SafeSuspense>
    </DefaultErrorBoundary>
  );
};
```

You can pass the query key(s) as the selector to specify queries that should be watched. There are some edge cases where this approach doesn't work though.
